C/STrace/EtwLogger.h

```diff
@@ -0,0 +1,314 @@
+{
+	UNREFERENCED_PARAMETER(fieldValue);
+
+	strcpy((char*)current, fieldName);
+	current[strlen(fieldName) + 1] = (uint8_t)fieldType;
+
+	current += strlen(fieldName) + 1 + sizeof(uint8_t);
+	SetFieldMetadata(current, rest...);
+}
+
+template<typename... Arguments>
+EVENT_DATA_DESCRIPTOR CreateEventMetadata(const char* eventName, Arguments... args)
+{
+	// Allocate the total size of the event metadata structure.
+	//
+	// This consists of, in order:
+	//
+	// * The total length of the structure
+	// * The event tag byte (currently always 0)
+	// * The name of the event
+	// * An array of field metadata structures, which are:
+	//     * the name of the field
+	//     * a single byte for the type.
+	const auto eventMetadataHeaderLength = strlen(eventName) + 1 + sizeof(uint16_t) + sizeof(uint8_t);
+	const auto eventMetadataLength = (uint16_t)(eventMetadataHeaderLength + SizeOfFieldsMeta(args...));
+	const auto eventMetadata = (struct detail::EventMetadata*)ExAllocatePoolWithTag(NonPagedPoolNx, eventMetadataLength, 'wteE');
+	if (eventMetadata == NULL)
+	{
+		return EVENT_DATA_DESCRIPTOR{};
+	}
+
+	RtlSecureZeroMemory(eventMetadata, eventMetadataLength);
+
+	// Set the first three fields, metadata about the event.
+	eventMetadata->TotalLength = eventMetadataLength;
+	eventMetadata->Tag = 0;
+	strcpy(eventMetadata->EventName, eventName);
+
+	// Set the metadata for each field.
+	uint8_t* currentLocation = ((uint8_t*)eventMetadata) + eventMetadataHeaderLength;
+	detail::SetFieldMetadata(currentLocation, args...);
+
+	// Create an EVENT_DATA_DESCRIPTOR pointing to the metadata.
+	EVENT_DATA_DESCRIPTOR eventMetadataDesc;
+	EventDataDescCreate(&eventMetadataDesc, eventMetadata, eventMetadata->TotalLength);
+	eventMetadataDesc.Type = EVENT_DATA_DESCRIPTOR_TYPE_EVENT_METADATA;  // Descriptor contains event metadata.
+
+	return eventMetadataDesc;
+}
+
+// final recursion case, string-specific overload
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, const char* fieldValue)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
+	UNREFERENCED_PARAMETER(fieldType);
+
+	// Copy the input value to its own space.
+	const auto newSpace = (char*)ExAllocatePoolWithTag(NonPagedPoolNx, strlen(fieldValue) + 1, 'wteE');
+	if (newSpace == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	strcpy(newSpace, fieldValue);
+
+	// Create the event data descriptor pointing to the value of the field.
+	EventDataDescCreate(OUT propertyDataDescriptors, newSpace, (ULONG)(strlen(fieldValue) + 1));
+
+	return STATUS_SUCCESS;
+}
+
+// final recursion case
+template<typename FieldValue>
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, FieldValue& fieldValue)
```

Comment:
Please don't use the OUT/IN meta tags, I understand microsoft does in their interfaces but in this project I would prefer not to use them. I would personally call this parameter pPropertyDataDescriptors which at least tells a caller it accepts a pointer and then they can make the assumption it's used as an out parameter.